File: app/src/main/java/com/kelsos/mbrc/adapters/PlaylistAdapter.java
package com.kelsos.mbrc.adapters;

import android.app.Activity;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.data.Playlist;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;

public class PlaylistAdapter extends RecyclerView.Adapter<PlaylistAdapter.ViewHolder> {

  private final LayoutInflater inflater;
  private List<Playlist> data;
  private OnPlaylistPressedListener playlistPressedListener;

  @Inject public PlaylistAdapter(Activity context) {
    inflater = LayoutInflater.from(context);
    data = new ArrayList<>();
  }

  @Override public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final View view = inflater.inflate(R.layout.listitem_single, parent, false);
    ViewHolder viewHolder = new ViewHolder(view);
    viewHolder.context.setVisibility(View.GONE);
    viewHolder.itemView.setOnClickListener(v -> {
      if (playlistPressedListener == null) {
        return;
      }
      playlistPressedListener.playlistPressed(data.get(viewHolder.getAdapterPosition()).getUrl());
    });
    return viewHolder;
  }

  @Override public void onBindViewHolder(ViewHolder holder, int position) {
    Playlist playlist = data.get(holder.getAdapterPosition());
    holder.name.setText(playlist.getName());
  }

  @Override public int getItemCount() {
    return data.size();
  }

  public void update(List<Playlist> playlist) {
    this.data.clear();

    if (playlist == null) {
      return;
    }

    this.data = playlist;
    notifyDataSetChanged();
  }

  public void setPlaylistPressedListener(OnPlaylistPressedListener playlistPressedListener) {
    this.playlistPressedListener = playlistPressedListener;
  }

  public interface OnPlaylistPressedListener {
    void playlistPressed(String path);
  }

  public static class ViewHolder extends RecyclerView.ViewHolder {
    @BindView(R.id.line_one) TextView name;
    @BindView(R.id.ui_item_context_indicator) LinearLayout context;

    public ViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/adapters/PlaylistAdapter.kt
package com.kelsos.mbrc.adapters

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.kelsos.mbrc.R
import com.kelsos.mbrc.domain.Playlist
import com.kelsos.mbrc.ui.navigation.playlists.tracks.PlaylistTrackActivity
import com.raizlabs.android.dbflow.list.FlowCursorList
import javax.inject.Inject

class PlaylistAdapter
@Inject constructor(context: Activity) : RecyclerView.Adapter<PlaylistAdapter.ViewHolder>() {

  private val inflater: LayoutInflater
  private var data: FlowCursorList<Playlist>? = null
  private var playlistPressedListener: OnPlaylistPressedListener? = null

  init {
    inflater = LayoutInflater.from(context)
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    val view = inflater.inflate(R.layout.listitem_single, parent, false)
    val viewHolder = ViewHolder(view)

    viewHolder.itemView.setOnClickListener { view: View? ->
      val playlist = data?.getItem(viewHolder.adapterPosition.toLong())
      playlist?.let {
        startPlaylistActivity(it, view)
      }

    }
    return viewHolder
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    val playlist = data?.getItem(holder.adapterPosition.toLong())
    playlist?.let {
      holder.name.text = playlist.name
    }
    holder.context.setOnClickListener {
      val menu = PopupMenu(it.context, it)
      menu.inflate(R.menu.popup_playlist)
      menu.show()

      menu.setOnMenuItemClickListener listener@ {
        if (it.itemId == R.id.playlist_tracks) {
          //startPlaylistActivity(playlist, it)
          return@listener true
        } else if (it.itemId == R.id.playlist_play) {
          val adapterPosition = holder.adapterPosition.toLong()
          val selectedPlaylist = data?.getItem(adapterPosition)
          //todo listener return
          return@listener true
        }
        false
      }
    }

  }

  override fun getItemCount(): Int {
    return data?.count ?: 0
  }

  private fun startPlaylistActivity(playlist: Playlist, v: View?) {
    if (v == null) {
      return
    }

    val bundle = Bundle()
    bundle.putString(PlaylistTrackActivity.NAME, playlist.name)
    bundle.putLong(PlaylistTrackActivity.ID, playlist.id)
    val intent = Intent(v.context, PlaylistTrackActivity::class.java)
    intent.putExtras(bundle)
    v.context.startActivity(intent)
  }

  fun update(cursor: FlowCursorList<Playlist>) {
    this.data = cursor
    notifyDataSetChanged()
  }

  fun setPlaylistPressedListener(playlistPressedListener: OnPlaylistPressedListener) {
    this.playlistPressedListener = playlistPressedListener
  }

  interface OnPlaylistPressedListener {
    fun playlistPressed(path: String)
  }

  class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    @BindView(R.id.line_one) lateinit var name: TextView
    @BindView(R.id.ui_item_context_indicator) lateinit var context: LinearLayout

    init {
      ButterKnife.bind(this, itemView)
    }
  }
}
